File: sdk-api/src/main/java/dev/restate/sdk/ContextImpl.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import dev.restate.common.*;
import dev.restate.common.function.ThrowingSupplier;
import dev.restate.sdk.common.DurablePromiseKey;
import dev.restate.sdk.common.HandlerRequest;
import dev.restate.sdk.common.RetryPolicy;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.endpoint.definition.AsyncResult;
import dev.restate.sdk.endpoint.definition.HandlerContext;
import dev.restate.serde.Serde;
import dev.restate.serde.SerdeFactory;
import dev.restate.serde.TypeTag;
import java.time.Duration;
import java.util.Collection;
import java.util.Collections;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

class ContextImpl implements ObjectContext, WorkflowContext {

  private static final ThreadLocal<Boolean> INSIDE_RUN = new ThreadLocal<>();

  private final HandlerContext handlerContext;
  private final Executor serviceExecutor;
  private final SerdeFactory serdeFactory;

  ContextImpl(HandlerContext handlerContext, Executor serviceExecutor, SerdeFactory serdeFactory) {
    this.handlerContext = handlerContext;
    this.serviceExecutor = serviceExecutor;
    this.serdeFactory = serdeFactory;
  }

  static void checkNotInsideRun() {
    if (Boolean.TRUE.equals(INSIDE_RUN.get())) {
      throw new IllegalStateException(
          "Cannot invoke context method inside ctx.run(). "
              + "The run closure is meant for non-deterministic operations (e.g., HTTP calls, database reads). "
              + "You MUST use context methods outside of ctx.run(), check the documentation: https://docs.restate.dev/develop/java/durable-steps#run");
    }
  }

  @Override
  public String key() {
    return handlerContext.objectKey();
  }

  @Override
  public HandlerRequest request() {
    return handlerContext.request();
  }

  @Override
  public <T> Optional<T> get(StateKey<T> key) {
    checkNotInsideRun();
    return DurableFuture.fromAsyncResult(
            Util.awaitCompletableFuture(handlerContext.get(key.name())), serviceExecutor)
        .mapWithoutExecutor(opt -> opt.map(serdeFactory.create(key.serdeInfo())::deserialize))
        .await();
  }

  @Override
  public Collection<String> stateKeys() {
    checkNotInsideRun();
    return Util.awaitCompletableFuture(
        Util.awaitCompletableFuture(handlerContext.getKeys()).poll());
  }

  @Override
  public void clear(StateKey<?> key) {
    checkNotInsideRun();
    Util.awaitCompletableFuture(handlerContext.clear(key.name()));
  }

  @Override
  public void clearAll() {
    checkNotInsideRun();
    Util.awaitCompletableFuture(handlerContext.clearAll());
  }

  @Override
  public <T> void set(StateKey<T> key, @NonNull T value) {
    checkNotInsideRun();
    Util.awaitCompletableFuture(
        handlerContext.set(
            key.name(),
            Util.executeOrFail(
                handlerContext, serdeFactory.create(key.serdeInfo())::serialize, value)));
  }

  @Override
  public DurableFuture<Void> timer(String name, Duration duration) {
    checkNotInsideRun();
    return DurableFuture.fromAsyncResult(
        Util.awaitCompletableFuture(handlerContext.timer(duration, name)), serviceExecutor);
  }

  @Override
  public <T, R> CallDurableFuture<R> call(Request<T, R> request) {
    checkNotInsideRun();
    Slice input =
        Util.executeOrFail(
            handlerContext,
            serdeFactory.create(request.getRequestTypeTag())::serialize,
            request.getRequest());
    HandlerContext.CallResult result =
        Util.awaitCompletableFuture(
            handlerContext.call(
                request.getTarget(),
                input,
                request.getIdempotencyKey(),
                request.getHeaders() == null
                    ? Collections.emptyList()
                    : request.getHeaders().entrySet()));

    return new CallDurableFuture<>(
        handlerContext,
        result
            .callAsyncResult()
            .map(
                s ->
                    CompletableFuture.completedFuture(
                        serdeFactory.create(request.getResponseTypeTag()).deserialize(s))),
        DurableFuture.fromAsyncResult(result.invocationIdAsyncResult(), serviceExecutor));
  }

  @Override
  public <Req, Res> InvocationHandle<Res> send(
      Request<Req, Res> request, @Nullable Duration delay) {
    checkNotInsideRun();
    Slice input =
        Util.executeOrFail(
            handlerContext,
            serdeFactory.create(request.getRequestTypeTag())::serialize,
            request.getRequest());

    var invocationIdAwaitable =
        DurableFuture.fromAsyncResult(
            Util.awaitCompletableFuture(
                handlerContext.send(
                    request.getTarget(),
                    input,
                    request.getIdempotencyKey(),
                    request.getHeaders() == null
                        ? Collections.emptyList()
                        : request.getHeaders().entrySet(),
                    delay)),
            serviceExecutor);

    return new BaseInvocationHandle<>(
        Util.executeOrFail(
            handlerContext, () -> serdeFactory.create(request.getResponseTypeTag()))) {
      @Override
      public String invocationId() {
        return invocationIdAwaitable.await();
      }
    };
  }

  @Override
  public <R> InvocationHandle<R> invocationHandle(String invocationId, TypeTag<R> responseTypeTag) {
    checkNotInsideRun();
    return new BaseInvocationHandle<>(
        Util.executeOrFail(handlerContext, () -> serdeFactory.create(responseTypeTag))) {
      @Override
      public String invocationId() {
        return invocationId;
      }
    };
  }

  abstract class BaseInvocationHandle<R> implements InvocationHandle<R> {
    private final Serde<R> responseSerde;

    BaseInvocationHandle(Serde<R> responseSerde) {
      this.responseSerde = responseSerde;
    }

    @Override
    public void cancel() {
      checkNotInsideRun();
      Util.awaitCompletableFuture(handlerContext.cancelInvocation(invocationId()));
    }

    @Override
    public DurableFuture<R> attach() {
      checkNotInsideRun();
      return DurableFuture.fromAsyncResult(
          Util.awaitCompletableFuture(handlerContext.attachInvocation(invocationId()))
              .map(s -> CompletableFuture.completedFuture(responseSerde.deserialize(s))),
          serviceExecutor);
    }

    @Override
    public Output<R> getOutput() {
      checkNotInsideRun();
      return DurableFuture.fromAsyncResult(
              Util.awaitCompletableFuture(handlerContext.getInvocationOutput(invocationId()))
                  .map(o -> CompletableFuture.completedFuture(o.map(responseSerde::deserialize))),
              serviceExecutor)
          .await();
    }
  }

  @Override
  public <T> DurableFuture<T> runAsync(
      String name, TypeTag<T> typeTag, RetryPolicy retryPolicy, ThrowingSupplier<T> action) {
    checkNotInsideRun();
    Serde<T> serde = serdeFactory.create(typeTag);
    return DurableFuture.fromAsyncResult(
            Util.awaitCompletableFuture(
                handlerContext.submitRun(
                    name,
                    runCompleter ->
                        serviceExecutor.execute(
                            () -> {
                              Slice result;
                              INSIDE_RUN.set(Boolean.TRUE);
                              try {
                                result = serde.serialize(action.get());
                              } catch (Throwable e) {
                                INSIDE_RUN.remove();
                                runCompleter.proposeFailure(e, retryPolicy);
                                return;
                              }
                              INSIDE_RUN.remove();
                              runCompleter.proposeSuccess(result);
                            }))),
            serviceExecutor)
        .mapWithoutExecutor(serde::deserialize);
  }

  @Override
  public <T> Awakeable<T> awakeable(TypeTag<T> typeTag) throws TerminalException {
    checkNotInsideRun();
    Serde<T> serde = serdeFactory.create(typeTag);
    // Retrieve the awakeable
    HandlerContext.Awakeable awakeable = Util.awaitCompletableFuture(handlerContext.awakeable());
    return new Awakeable<>(awakeable.asyncResult(), serviceExecutor, serde, awakeable.id());
  }

  @Override
  public AwakeableHandle awakeableHandle(String id) {
    return new AwakeableHandle() {
      @Override
      public <T> void resolve(TypeTag<T> serde, @NonNull T payload) {
        checkNotInsideRun();
        Util.awaitCompletableFuture(
            handlerContext.resolveAwakeable(
                id,
                Util.executeOrFail(
                    handlerContext, serdeFactory.create(serde)::serialize, payload)));
      }

      @Override
      public void reject(String reason) {
        checkNotInsideRun();
        Util.awaitCompletableFuture(
            handlerContext.rejectAwakeable(id, new TerminalException(reason)));
      }
    };
  }

  @Override
  public RestateRandom random() {
    return new RestateRandom(this.request().invocationId().toRandomSeed());
  }

  @Override
  public <T> DurablePromise<T> promise(DurablePromiseKey<T> key) {
    return new DurablePromise<>() {
      @Override
      public DurableFuture<T> future() {
        checkNotInsideRun();
        AsyncResult<Slice> result = Util.awaitCompletableFuture(handlerContext.promise(key.name()));
        return DurableFuture.fromAsyncResult(result, serviceExecutor)
            .mapWithoutExecutor(serdeFactory.create(key.serdeInfo())::deserialize);
      }

      @Override
      public Output<T> peek() {
        checkNotInsideRun();
        return Util.awaitCompletableFuture(
                Util.awaitCompletableFuture(handlerContext.peekPromise(key.name())).poll())
            .map(serdeFactory.create(key.serdeInfo())::deserialize);
      }
    };
  }

  @Override
  public <T> DurablePromiseHandle<T> promiseHandle(DurablePromiseKey<T> key) {
    return new DurablePromiseHandle<>() {
      @Override
      public void resolve(T payload) throws IllegalStateException {
        checkNotInsideRun();
        Util.awaitCompletableFuture(
            Util.awaitCompletableFuture(
                    handlerContext.resolvePromise(
                        key.name(),
                        Util.executeOrFail(
                            handlerContext,
                            serdeFactory.create(key.serdeInfo())::serialize,
                            payload)))
                .poll());
      }

      @Override
      public void reject(String reason) throws IllegalStateException {
        checkNotInsideRun();
        Util.awaitCompletableFuture(
            Util.awaitCompletableFuture(
                    handlerContext.rejectPromise(key.name(), new TerminalException(reason)))
                .poll());
      }
    };
  }
}

File: sdk-api-kotlin/src/main/kotlin/dev/restate/sdk/kotlin/ContextImpl.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import dev.restate.common.Output
import dev.restate.common.Request
import dev.restate.common.SendRequest
import dev.restate.common.Slice
import dev.restate.sdk.endpoint.definition.HandlerContext
import dev.restate.sdk.types.DurablePromiseKey
import dev.restate.sdk.types.HandlerRequest
import dev.restate.sdk.types.StateKey
import dev.restate.sdk.types.TerminalException
import dev.restate.serde.Serde
import dev.restate.serde.SerdeFactory
import dev.restate.serde.TypeTag
import java.util.concurrent.CompletableFuture
import kotlin.jvm.optionals.getOrNull
import kotlin.time.Duration
import kotlin.time.toJavaDuration
import kotlinx.coroutines.*
import kotlinx.coroutines.future.await

internal class ContextImpl
internal constructor(
    internal val handlerContext: HandlerContext,
    internal val contextSerdeFactory: SerdeFactory
) : WorkflowContext {
  override fun key(): String {
    return this.handlerContext.objectKey()
  }

  override fun request(): HandlerRequest {
    return this.handlerContext.request()
  }

  override suspend fun <T : Any> get(key: StateKey<T>): T? =
      resolveSerde<T?>(key.serdeInfo())
          .let { serde ->
            SingleAwaitableImpl(handlerContext.get(key.name()).await()).simpleMap {
              it.getOrNull()?.let { serde.deserialize(it) }
            }
          }
          .await()

  override suspend fun stateKeys(): Collection<String> =
      SingleAwaitableImpl(handlerContext.getKeys().await()).await()

  override suspend fun <T : Any> set(key: StateKey<T>, value: T) {
    handlerContext.set(key.name(), resolveAndSerialize(key.serdeInfo(), value)).await()
  }

  override suspend fun clear(key: StateKey<*>) {
    handlerContext.clear(key.name()).await()
  }

  override suspend fun clearAll() {
    handlerContext.clearAll().await()
  }

  override suspend fun timer(duration: Duration, name: String?): Awaitable<Unit> =
      SingleAwaitableImpl(handlerContext.timer(duration.toJavaDuration(), name).await()).map {}

  override suspend fun <Req : Any?, Res : Any?> call(
      request: Request<Req, Res>
  ): CallAwaitable<Res> =
      resolveSerde<Res>(request.responseTypeTag()).let { responseSerde ->
        val callHandle =
            handlerContext
                .call(
                    request.target(),
                    resolveAndSerialize<Req>(request.requestTypeTag(), request.request()),
                    request.idempotencyKey(),
                    request.headers().entries)
                .await()

        val callAsyncResult =
            callHandle.callAsyncResult.map {
              CompletableFuture.completedFuture<Res>(responseSerde.deserialize(it))
            }

        return@let CallAwaitableImpl(callAsyncResult, callHandle.invocationIdAsyncResult)
      }

  override suspend fun <Req : Any?, Res : Any?> send(
      request: Request<Req, Res>
  ): InvocationHandle<Res> =
      resolveSerde<Res>(request.responseTypeTag()).let { responseSerde ->
        val invocationIdAsyncResult =
            handlerContext
                .send(
                    request.target(),
                    resolveAndSerialize<Req>(request.requestTypeTag(), request.request()),
                    request.idempotencyKey(),
                    request.headers().entries,
                    (request as? SendRequest)?.delay())
                .await()

        object : BaseInvocationHandle<Res>(handlerContext, responseSerde) {
          override suspend fun invocationId(): String = invocationIdAsyncResult.poll().await()
        }
      }

  override fun <Res> invocationHandle(
      invocationId: String,
      responseTypeTag: TypeTag<Res>
  ): InvocationHandle<Res> =
      resolveSerde<Res>(responseTypeTag).let { responseSerde ->
        object : BaseInvocationHandle<Res>(handlerContext, responseSerde) {
          override suspend fun invocationId(): String = invocationId
        }
      }

  override suspend fun <T : Any?> runAsync(
      typeTag: TypeTag<T>,
      name: String,
      retryPolicy: RetryPolicy?,
      block: suspend () -> T
  ): Awaitable<T> {
    var serde: Serde<T> = resolveSerde(typeTag)
    var coroutineCtx = currentCoroutineContext()
    val javaRetryPolicy =
        retryPolicy?.let {
          dev.restate.sdk.types.RetryPolicy.exponential(
                  it.initialDelay.toJavaDuration(), it.exponentiationFactor)
              .setMaxAttempts(it.maxAttempts)
              .setMaxDelay(it.maxDelay?.toJavaDuration())
              .setMaxDuration(it.maxDuration?.toJavaDuration())
        }

    val scope = CoroutineScope(coroutineCtx + CoroutineName("restate-run-$name"))

    val asyncResult =
        handlerContext
            .submitRun(name) { completer ->
              scope.launch {
                val result: Slice?
                try {
                  result = serde.serialize(block())
                } catch (e: Throwable) {
                  completer.proposeFailure(e, javaRetryPolicy)
                  return@launch
                }
                completer.proposeSuccess(result)
              }
            }
            .await()
    return SingleAwaitableImpl(asyncResult).map { serde.deserialize(it) }
  }

  override suspend fun <T : Any> awakeable(typeTag: TypeTag<T>): Awakeable<T> {
    val serde: Serde<T> = resolveSerde(typeTag)
    val awk = handlerContext.awakeable().await()
    return AwakeableImpl(awk.asyncResult, serde, awk.id)
  }

  override fun awakeableHandle(id: String): AwakeableHandle {
    return AwakeableHandleImpl(this, id)
  }

  override fun random(): RestateRandom {
    return RestateRandom(handlerContext.request().invocationId().toRandomSeed())
  }

  override fun <T : Any> promise(key: DurablePromiseKey<T>): DurablePromise<T> {
    return DurablePromiseImpl(key)
  }

  override fun <T : Any> promiseHandle(key: DurablePromiseKey<T>): DurablePromiseHandle<T> {
    return DurablePromiseHandleImpl(key)
  }

  inner class DurablePromiseImpl<T : Any>(private val key: DurablePromiseKey<T>) :
      DurablePromise<T> {
    val serde: Serde<T> = resolveSerde(key.serdeInfo())

    override suspend fun awaitable(): Awaitable<T> =
        SingleAwaitableImpl(handlerContext.promise(key.name()).await()).simpleMap {
          serde.deserialize(it)
        }

    override suspend fun peek(): Output<T> =
        SingleAwaitableImpl(handlerContext.peekPromise(key.name()).await())
            .simpleMap { it.map { serde.deserialize(it) } }
            .await()
  }

  inner class DurablePromiseHandleImpl<T : Any>(private val key: DurablePromiseKey<T>) :
      DurablePromiseHandle<T> {
    val serde: Serde<T> = resolveSerde(key.serdeInfo())

    override suspend fun resolve(payload: T) {
      SingleAwaitableImpl(
              handlerContext
                  .resolvePromise(
                      key.name(), serde.serializeWrappingException(handlerContext, payload))
                  .await())
          .await()
    }

    override suspend fun reject(reason: String) {
      SingleAwaitableImpl(
              handlerContext.rejectPromise(key.name(), TerminalException(reason)).await())
          .await()
    }
  }

  internal fun <T : Any?> resolveAndSerialize(typeTag: TypeTag<T>, value: T): Slice {
    return try {
      val serde = contextSerdeFactory.create<T>(typeTag)
      serde.serialize(value)
    } catch (e: Exception) {
      handlerContext.fail(e)
      throw CancellationException("Failed serialization", e)
    }
  }

  private fun <T : Any?> resolveSerde(typeTag: TypeTag<T>): Serde<T> {
    return try {
      contextSerdeFactory.create<T>(typeTag)!!
    } catch (e: Exception) {
      handlerContext.fail(e)
      throw CancellationException("Cannot resolve serde", e)
    }
  }
}
